funcframework/events.go

```diff
@@ -83,20 +85,34 @@ var (
 	}
 )
 
-type backgroundEvent struct {
-	Data     interface{}        `json:"data"`
-	Metadata *metadata.Metadata `json:"context"`
-}
+func getBackgroundEvent(body []byte, path string) (*metadata.Metadata, interface{}, error) {
+	// Known background event types that the incoming request could represent.
+	// Event types are mutually exclusive. During unmarshalling, only the field
+	// for the matching type is populated.
+	type possibleEvents struct {
+		*pubsub.LegacyEvent
+		*fftypes.BackgroundEvent
+	}
 
-func getBackgroundEvent(body []byte) (*metadata.Metadata, interface{}, error) {
-	// Handle background events' "data" and "context" fields.
-	event := backgroundEvent{}
-	if err := json.Unmarshal(body, &event); err != nil {
+	// Attempt to unmarshal into one of the known background event types.
+	possible := possibleEvents{}
+	if err := json.Unmarshal(body, &possible); err != nil {
 		return nil, nil, err
 	}
 
+	event := possible.BackgroundEvent
+	// If the background event payload is missing, check if it's a legacy
+	// Pub/Sub event.
+	if possible.BackgroundEvent == nil && possible.LegacyEvent != nil {
```

Comment:
nit: I think `LegacyEvent` isn't a great name for this because, at this point, it isn't clear that this logic is specific to PubSub. Maybe give the fields in the `possibleEvents` struct names?

Reply:
There wasn't a great way to include fieldnames because the way the `json` library works, it automatically expects field names to be present in the JSON. But at least the new "LegacyPushSubscriptionEvent" field name heavily implies that it's a Pub/Sub related.